From 558145909e641ed58257634501a4b494567499c2 Mon Sep 17 00:00:00 2001
From: Quanyang Wang <quanyang.wang@windriver.com>
Date: Sun, 27 Jan 2019 20:13:25 +0800
Subject: [PATCH] genirq/irqdesc: Fix spinlock bad magic issue

In upstream commit 9114014cf4e6 ("genirq: Add mutex to irq desc to serialize request/free_irq()"),
it introduces desc->request_mutex, but in some case it isn't initialised. So init it
to avoid the calltrace as below:

[    0.000000] BUG: spinlock bad magic on CPU#0, swapper/0/0
[    0.000000]  lock: irq_desc+0x3b40/0x20000, .magic: 00000000, .owner: swapper/0/0, .owner_cpu: 0
[    0.000000] CPU: 0 PID: 0 Comm: swapper/0 Not tainted 4.18.20-rt8-yocto-preempt-rt+ #157
[    0.000000] Hardware name: INTEL Axxia
[    0.000000] [<c06155b8>] (unwind_backtrace) from [<c060e6ac>] (show_stack+0x20/0x24)
[    0.000000] [<c060e6ac>] (show_stack) from [<c10d5de0>] (dump_stack+0x9c/0xd0)
[    0.000000] [<c10d5de0>] (dump_stack) from [<c06a125c>] (spin_bug+0x9c/0xb0)
[    0.000000] [<c06a125c>] (spin_bug) from [<c06a13fc>] (do_raw_spin_unlock+0x30/0xdc)
[    0.000000] [<c06a13fc>] (do_raw_spin_unlock) from [<c10f0ebc>] (_raw_spin_unlock_irqrestore+0x34/0x90)
[    0.000000] [<c10f0ebc>] (_raw_spin_unlock_irqrestore) from [<c10eed18>] (rt_mutex_slowlock.constprop.11+0x7c/0xb8)
[    0.000000] [<c10eed18>] (rt_mutex_slowlock.constprop.11) from [<c10eede8>] (__rt_mutex_lock_state+0x94/0x9c)
[    0.000000] [<c10eede8>] (__rt_mutex_lock_state) from [<c10f1458>] (_mutex_lock+0x48/0x50)
[    0.000000] [<c10f1458>] (_mutex_lock) from [<c06afc34>] (__setup_irq+0x1a4/0x72c)
[    0.000000] [<c06afc34>] (__setup_irq) from [<c06b04e8>] (__request_percpu_irq+0xd0/0xf0)
[    0.000000] [<c06b04e8>] (__request_percpu_irq) from [<c1655f7c>] (arch_timer_of_init+0x200/0x314)
[    0.000000] [<c1655f7c>] (arch_timer_of_init) from [<c16559e8>] (timer_probe+0x94/0xe0)
[    0.000000] [<c16559e8>] (timer_probe) from [<c160c1d8>] (axxia_dt_timer_init+0x3c/0x44)
[    0.000000] [<c160c1d8>] (axxia_dt_timer_init) from [<c16064dc>] (time_init+0x2c/0x40)
[    0.000000] [<c16064dc>] (time_init) from [<c1600e64>] (start_kernel+0x300/0x4e4)
[    0.000000] [<c1600e64>] (start_kernel) from [<00000000>] (  (null))

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 kernel/irq/irqdesc.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/kernel/irq/irqdesc.c b/kernel/irq/irqdesc.c
index afc7f90..1a7e806 100644
--- a/kernel/irq/irqdesc.c
+++ b/kernel/irq/irqdesc.c
@@ -553,6 +553,7 @@ int __init early_irq_init(void)
 		raw_spin_lock_init(&desc[i].lock);
 		lockdep_set_class(&desc[i].lock, &irq_desc_lock_class);
 		desc_set_defaults(i, &desc[i], node, NULL, NULL);
+		mutex_init(&desc[i].request_mutex);
 	}
 	return arch_early_irq_init();
 }
-- 
1.7.5.4

